dsl_parser/functions.py

```diff
@@ -850,63 +921,36 @@ def runtime_evaluation_handler(get_node_instances_method,
 
 
 def validate_functions(plan):
-    get_property_functions = []
+    level_of_tree_has_runtime_func = [False]
 
     def handler(v, scope, context, path):
-        _func = parse(v, scope=scope, context=context, path=path)
-        if isinstance(_func, Function):
-            _func.validate(plan)
-        if isinstance(_func, GetProperty):
-            get_property_functions.append(_func)
-            return _func
+        func = parse(v, scope=scope, context=context, path=path)
+        if isinstance(func, Function):
+            func.validate(plan)
+            level_of_tree_has_runtime_func.append(False)
+        scan.scan_properties(
```

Comment:
this is essentially a depth-first search i think, but it's very complex, also due to how we keep a whole list of the `level_of_tree_has_runtime_func`? one bool should be enough i think, no?

i think it ends up being one bool anyway due to the `.pop` but we somewhat hide that; maybe we can replace this with just a `level[0] = level[0] or func.eval_type == RUNTIME` etc. (using a list with one element, which is a common idiom in python 2 with no `nonlocal`)
(or we could write a class and have the handler here be a method, and the bool be an instance attribute)

either way, this function could sure use some comments

Reply:
(Continued on Slack)